### PR TITLE
feat(typescript): start ts set up

### PR DESCRIPTION
**Pull request by ferdicus**
This is supposed to be the initial set up for integrating Typescript into the project.
The idea is to change the project bit by bit, instead of having a huge refactoring up front.
Having said that, this can't be merged just yet, because some things are missing, however I wanted to get some eyes on this before I continue.

The big idea would be to, before we publish, run the build script and have a `lib/` directory with all compiled `ts` => `js` files and a nice `index.d.ts` generated from our (to be created) `types` folder.

What's in here so far:
* `tsconfig`, which grabs all `.ts` files within `<root>/javascript/` and puts them in `<root>/lib/`
   currently this is only one small, simple file which I converted: `javascript/utils/filterUtils.ts `
* till we convert all `.js` files to `.ts` files, we need to copy `.js` files in `<root>/lib/`, that's why I added `cpy`, which copies all non `.ts` files from `<root>/javascript/` over to `<root>/lib/`

What's missing:
* no `types` 😞 (yet 😄 )
* no `index.d.ts` generated out of our `types`
* haven't run tests
* ~haven't tested with `example/`~

Looking forward to discussions/ conversations about this.

Cheers 🍺 

-------

pays into #921

### CODE REVIEW

File: javascript/utils/filterUtils.ts [hunk 1]
```diff
@@ -1,4 +1,5 @@
-export function getFilter(filter) {
+export function getFilter(filter: any) {
+  console.log('>>>>>> ');
```
Comment (ClayC-WeatherBug) [inline javascript/utils/filterUtils.ts:2]: Is this console logging statement still needed?

**Review by ClayC-WeatherBug — APPROVED** [review]

File: javascript/utils/filterUtils.ts (hunk 1 above)
Comment (ferdicus) [reply javascript/utils/filterUtils.ts:2]: not at all - I just wanted to make sure that the file had some tangible change, that I can see 👀 in the logs :D

**Comment by ferdicus** [comment]
hey @mfazekas, do you think this is a good first step - are you fine with the direction and can I merge this?

Cheers 🍻

**Comment by mfazekas** [comment]
@ferdicus I think the direction is fine.

I have a few notes/questions/concerns:

- copying of the files can be problematic as it means we can no longer just edit the source, we'll need to ran the copy, too. It would be great if we could work with the example app, by just editing library files. We might need some metro magic for that?!
- for class hierarchies it's not clear if we can do gradually, and if we can/should convert from top or from bottom.  So Light or AbstractLayer might be an ideal candidate for a bit more complex case.

**Comment by ferdicus** [comment]
> copying of the files can be problematic as it means we can no longer just edit the source, we'll need to ran the copy, too. 

I just tested - it works, no need for copy 👍

However, I can only confirm it for regular refresh of the app, as I've noticed, that we seem to have a `fast reload` bug with `/example` when we alter files in `<root>`.  This does not apply to files edited in `/example`.


With `fast refresh` it complains like so:
```sh
None of these files exist:
  * javascript/utils/filterUtils(.native|.ios.js|.native.js|.js|.ios.json|.native.json|.json|.ios.ts|.native.ts|.ts|.ios.tsx|.native.tsx|.tsx)
  * javascript/utils/filterUtils/index(.native|.ios.js|.native.js|.js|.ios.json|.native.json|.json|.ios.ts|.native.ts|.ts|.ios.tsx|.native.tsx|.tsx)
    at ModuleResolver.resolveDependency (/Users/john/Development/Maps/example/node_modules/metro/src/node-haste/DependencyGraph/ModuleResolution.js:163:15)
    at ResolutionRequest.resolveDependency (/Users/john/Development/Maps/example/node_modules/metro/src/node-haste/DependencyGraph/ResolutionRequest.js:52:18)
    at DependencyGraph.resolveDependency (/Users/john/Development/Maps/example/node_modules/metro/src/node-haste/DependencyGraph.js:287:16)
    at /Users/john/Development/Maps/example/node_modules/metro/src/lib/transformHelpers.js:267:42
    at /Users/john/Development/Maps/example/node_modules/metro/src/Server.js:1096:37
    at Generator.next (<anonymous>)
    at asyncGeneratorStep (/Users/john/Development/Maps/example/node_modules/metro/src/Server.js:99:24)
    at _next (/Users/john/Development/Maps/example/node_modules/metro/src/Server.js:119:9)
    at process._tickCallback (internal/process/next_tick.js:68:7)
[Fri Oct 16 2020 18:16:36.982]  ERROR    Invariant Violation: Tried to register two views with the same name RCTMGLShapeSource
```

I couldn't quickly figure out what the issue was unfortunately

**Comment by RichardLindhout** [comment]
https://github.com/callstack/react-native-bob Could be great for setting up the project

**Comment by ferdicus** [comment]
> https://github.com/callstack/react-native-bob Could be great for setting up the project

Hmm, I don't see how this is helping in the current state of the TS progress - before I started and had the current setup, yes, this would've made sense I guess. Now it would maybe add more hassle/ overhead instead 😅 

Thinks we have to do now is the heavy lifting of actually rewriting the JS code to TS 😓

**Comment by ClayC-WeatherBug** [comment]
> > https://github.com/callstack/react-native-bob Could be great for setting up the project
> 
> Hmm, I don't see how this is helping in the current state of the TS progress - before I started and had the current setup, yes, this would've made sense I guess. Now it would maybe add more hassle/ overhead instead 😅
> 
> Thinks we have to do now is the heavy lifting of actually rewriting the JS code to TS 😓

As far as converting goes, there is a VSCode extension that'll auto-convert components to TS. If they use prop types it converts those to an interface. It's not perfect, and won't help with places where we need to define the types and things like that, but it may help nudge things along a bit faster than doing it manually.

Extension: https://marketplace.visualstudio.com/items?itemName=mohsen1.react-javascript-to-typescript-transform-vscode

**Comment by ferdicus** [comment]
anyone feel free to chime in here - what I find super weird:
* I've converted `UserLocation` from js => tsx
* I did not compile anything afterwards 
* pointed a newly initialized `react-native init` project to this last commit, with `UserLocation.tsx`
* build the app for android with a very simple map, showing the users location - for shits and giggles, expecting it to crap out
* it actually build, and shows the users location.... 🤔 😕 🤔 

any ideas? is react-native nowadays supporting TS for `node_modules` out of the box?!

`package.json` of test project:
```json
{
  "name": "test",
  "version": "0.0.1",
  "private": true,
  "scripts": {
    "android": "react-native run-android",
    "ios": "react-native run-ios",
    "start": "react-native start",
    "test": "jest",
    "lint": "eslint ."
  },
  "dependencies": {
    "react": "16.13.1",
    "react-native": "0.63.4"
  },
  "devDependencies": {
    "@babel/core": "^7.8.4",
    "@babel/runtime": "^7.8.4",
    "@react-native-community/eslint-config": "^1.1.0",
    "@react-native-mapbox-gl/maps": "git://github.com/react-native-mapbox-gl/maps.git#a2a38dabe238984a7c27b8c236a45b014f01c820",
    "babel-jest": "^25.1.0",
    "eslint": "^6.5.1",
    "jest": "^25.1.0",
    "metro-react-native-babel-preset": "^0.59.0",
    "react-test-renderer": "16.13.1"
  },
  "jest": {
    "preset": "react-native"
  }
}
```

that is as clean as a react-native project can be....

**Comment by ClayC-WeatherBug** [comment]
> * I did not compile anything afterwards

@ferdicus if you didn't compile anything after making this change then I would expect that it would be running the code prior to you converting the component to TS. If you did actually run `yarn build`, then it looks like this PR switches over to using the typescript compiler so I would expect it to also work, although our tsconfig only has `*.ts` and not `*.tsx` which I'm not sure if it will pick up.

**Comment by ferdicus** [comment]
> > * I did not compile anything afterwards
> 
> @ferdicus if you didn't compile anything after making this change then I would expect that it would be running the code prior to you converting the component to TS. If you did actually run `yarn build`, then it looks like this PR switches over to using the typescript compiler so I would expect it to also work, although our tsconfig only has `*.ts` and not `*.tsx` which I'm not sure if it will pick up.

Thanks for the input @ClayC-WeatherBug, 
yeah it's super weird:
* I didn't build
* `UserLocation.js` was _replaced_ by `UserLocation.tsx`

🤷🏿

**Comment by ClayC-WeatherBug** [comment]
> > > * I did not compile anything afterwards
> > 
> > 
> > @ferdicus if you didn't compile anything after making this change then I would expect that it would be running the code prior to you converting the component to TS. If you did actually run `yarn build`, then it looks like this PR switches over to using the typescript compiler so I would expect it to also work, although our tsconfig only has `*.ts` and not `*.tsx` which I'm not sure if it will pick up.
> 
> Thanks for the input @ClayC-WeatherBug,
> yeah it's super weird:
> 
> * I didn't build
> * `UserLocation.js` was _replaced_ by `UserLocation.tsx`
> 
> 🤷🏿

Hm, that is a bit weird. I haven't looked too closely at our build setup, but I would expect that the .tsx file would have been replaced, but that the actual build that we output to still be the same old build. Don't we bundle up the JS into some kind of `dist.js` file or equivalent as part of the build step?

**Review by ClayC-WeatherBug — APPROVED** [review]

**Comment by janicduplessis** [comment]
@ferdicus What is missing here? Can I help pushing this forward?

React native recently removed PropTypes so I was thinking this could be good opportunity to also migrate to typescript to replace PropTypes.

**Pull request closed without merging**